hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/common/table/read/TestDeleteRecordLogic.scala

```diff
@@ -0,0 +1,224 @@
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.{Arguments, MethodSource}
+
+class TestDeleteRecordLogic extends SparkClientFunctionalTestHarness{
+  val expected1 = Seq(
+    (14, "5", "rider-Z", "driver-Z", 17.85, 3),
+    (10, "3", "rider-C", "driver-C", 33.9, 10),
+    (10, "2", "rider-B", "driver-B", 27.7, 1))
+
+  val expected2 = Seq(
+    (14, "5", "rider-Z", "driver-Z", 17.85, 3),
+    (-9, "4", "rider-DDDD", "driver-DDDD", 20.0, 1),
+    (10, "3", "rider-C", "driver-C", 33.9, 10),
+    (10, "2", "rider-B", "driver-B", 27.7, 1))
+
+  @ParameterizedTest
+  @MethodSource(Array("provideParams"))
+  def testDeleteLogic(useFgReader: String, tableType: String, recordType: String): Unit = {
+    val sparkOpts: Map[String, String] = Map(
+      HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key -> "parquet",
+      HoodieWriteConfig.RECORD_MERGER_IMPLS.key -> classOf[DefaultSparkRecordMerger].getName)
+    val fgReaderOpts: Map[String, String] = Map(
+      HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key -> useFgReader)
+
+    val opts = if (recordType.equals("SPARK")) sparkOpts ++ fgReaderOpts else fgReaderOpts
+    val columns = Seq("ts", "key", "rider", "driver", "fare", "number")
+
+    val data = Seq(
+      (10, "1", "rider-A", "driver-A", 19.10, 7),
+      (10, "2", "rider-B", "driver-B", 27.70, 1),
+      (10, "3", "rider-C", "driver-C", 33.90, 10),
+      (-1, "4", "rider-D", "driver-D", 34.15, 6),
+      (10, "5", "rider-E", "driver-E", 17.85, 10))
+    val inserts = spark.createDataFrame(data).toDF(columns: _*)
+    inserts.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(DataSourceWriteOptions.TABLE_NAME.key(), "test_table").
+      option(HoodieCompactionConfig.INLINE_COMPACT.key(), "false").
+      options(opts).
+      mode(SaveMode.Overwrite).
+      save(basePath)
+
+    val updateData = Seq(
+      (11, "1", "rider-X", "driver-X", 19.10, 9),
+      (9, "2", "rider-Y", "driver-Y", 27.70, 7))
+    val updates = spark.createDataFrame(updateData).toDF(columns: _*)
+    updates.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "upsert").
+      option(HoodieCompactionConfig.INLINE_COMPACT.key(), "false").
+      options(opts).
+      mode(SaveMode.Append).
+      save(basePath)
+
+    val deletesData = Seq((-5, "4", "rider-D", "driver-D", 34.15, 6))
+    val deletes = spark.createDataFrame(deletesData).toDF(columns: _*)
+    deletes.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "delete").
+      option(HoodieClusteringConfig.INLINE_CLUSTERING.key(), "true").
```

Comment:
why we need to trigger clustering? if we trigger clustering, then there will be only base files. can we remove clustering from here

Reply:
Just tried to be more conprehensive. Ok, I will remove clutering.